SchemaStore fix-ups

chrisvest opened the pull request:
- [x] Add missing support for creating composite indexes through the Core API.
- [x] Fix serialisation of schema records that take up more than one dynamic record.
- [x] Add computed unique names to SchemaRule.
- [x] Add a name field to the new schema records for indexes and constraints, so that they in the future can be addressed by name on the surface – this will be used instead of the computed SchemaRule name, if it exists (and can initially be set to the computed SchemaRule name).

fickludd committed:
Can create composite index in Core API

chrisvest committed:
Fix test failure for creating composite indexes

chrisvest committed:
Fix serialisation of schema records that spill into multiple dynamic records

chrisvest committed:
Fix REST tests for composite indexes

chrisvest committed:
Add test that assert that REST cannot create composite constraints

chrisvest committed:
Support named rules in SchemaStore

- This adds an "optional" name field to the end of schema rules in the schema
  store.
- SchemaRule objects get a generated name if none is otherwise provided.
- The generated name is frozen once observed and cannot be changed afterwards.
- The schema rule names are thus effectively immutable.
- Legacy schema rule records cannot fit this new name anywhere, and thus always
  gets a generated name.
- New schema rule records that don't have a name gets a generated one as well.
- The empty string is not a valid name, but if we somehow find the empty string
  in the store, we will use the generated name instead – this allows us to add
  other fields in the future, without imposing a hard requirement that all
  future schema rule records must have a name.
- The name is stored as a null-terminated string, which is different from other
  strings in the schema store.
- The reason for the null-terminated strings is that its slightly more straight
  forward to handle if the name field is missing.
- Compatibility tests is added for all these new cases.
- Also, obviously, if you set an explicit schema rule name, we require it to
  not be the empty string, and to not contain the null-byte character.
- The name field string is encoded in UTF-8, so the terminating null-byte will
  be the only null-byte in the string.

ragadeeshu commented on community/kernel/src/test/java/org/neo4j/test/mockito/matcher/Neo4jMatchers.java:
```diff
@@ -618,20 +619,25 @@ public void describeTo( Description description )
     }
 
-    public static IndexDefinition createIndex( GraphDatabaseService beansAPI, Label label, String property )
+    public static IndexDefinition createIndex( GraphDatabaseService beansAPI, Label label, String... property )
```
`String... properties` to match `createIndexNoWait`?

ragadeeshu commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/SchemaStorageTest.java:
```diff
@@ -115,6 +119,109 @@ public void shouldReturnIndexRuleForLabelAndProperty()
+        assertEquals( NewIndexDescriptor.Type.GENERAL, rule.getIndexDescriptor().type() );
+    }
+
+    @Ignore("it is currently disputed whether constraints should be supported via Core API or not.")
```
Wow

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/RecordSerializer.java:
```diff
@@ -25,7 +25,7 @@
-    private final List<RecordSerializable> serializables = new ArrayList<RecordSerializable>();
+    private final List<RecordSerializable> serializables = new ArrayList<>();
 
     public RecordSerializer append( RecordSerializable serializable )
```
I think we could remove this class completely, but let's wait. Good simplifications.

fickludd commented on community/kernel/src/main/java/org/neo4j/storageengine/api/schema/SchemaRule.java:
```diff
@@ -26,23 +26,97 @@
+     * it cannot be changed.
+     * @param name The name desired for this rule.
+     * @return {@code true} if the name was set, {@code false} if this rule already has a name.
+     */
```
I don't know, I find this to be a bit too complicated? How about making the name a final field instead, and pass it into the constructor of the SchemaRules. Then we can make a factory method without the name, which would auto-generate the name. This would also avoid the instance of checks in `generateName`.

fickludd commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/SchemaStorageTest.java:
```diff
@@ -115,6 +119,109 @@ public void shouldReturnIndexRuleForLabelAndProperty()
+              .assertPropertyIsUnique( d ).assertPropertyIsUnique( e ).assertPropertyIsUnique( f ).create();
+        } );
+
+        IndexRule rule = storage.indexGetForSchema( NewIndexDescriptorFactory.forLabel(
```
this should be `NewIndexDescriptorFactory.uniqueForLabel` I think

fickludd commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/SchemaStorageTest.java:
```diff
@@ -115,6 +119,109 @@ public void shouldReturnIndexRuleForLabelAndProperty()
+        assertTrue( SchemaDescriptorPredicates.hasProperty( rule, propId( d ) ) );
+        assertTrue( SchemaDescriptorPredicates.hasProperty( rule, propId( e ) ) );
+        assertTrue( SchemaDescriptorPredicates.hasProperty( rule, propId( f ) ) );
+        assertEquals( NewIndexDescriptor.Type.UNIQUE, rule.getIndexDescriptor().type() );
```
Also you can do assert directly on `NewIndexDescriptor` equality if you don't care to see exactly what was wrong. Maybe we do in this case.

fickludd commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/record/SchemaRuleSerializationTest.java:
```diff
@@ -131,47 +230,75 @@ public void shouldReturnCorrectLengthForConstraintRules() throws MalformedSchema
+        assertParseRelationshipPropertyExistsRule( "/////ww9XAAAIUAAAQAAF+c=", "constraint_51" );
+        assertParseRelationshipPropertyExistsRule( "AAAhQAUAABfn", "constraint_51" ); // LEGACY6
+        assertParseRelationshipPropertyExistsRule( "/////ww9XAAAIUAAAQAAF+djdXN0b21fbmFtZQA=", "custom_name" ); // named rule
+        assertParseRelationshipPropertyExistsRule( addNullByte( "/////ww9XAAAIUAAAQAAF+c=" ), "constraint_51" ); // empty name
```
Lot's of testing, nice!

fickludd reviewed (commented):
I like it, let's just iterate the schema name implementation, and then I'm happy.

chrisvest committed:
Simplify SchemaRule serialisation.

Removes RecordSerializer and RecordSerializable, and makes serialisation a straight forward method call, that also allocates its own buffer.

chrisvest committed:
Remove ignored tests for functionality that it does not appear we want

chrisvest committed:
Make the SchemaRule name field final

chrisvest committed:
Fix typo

chrisvest commented:
I want to remove the null-terminated strings as well, and use the regular length-prefixed strings instead.

chrisvest committed:
Add test cases to CompositeIndexingIT for schema rules that span more than one dynamic record

chrisvest committed:
Remove the use of null-terminated strings in SchemaRule records

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/ConstraintRule.java:
```diff
@@ -43,9 +43,26 @@ public static ConstraintRule constraintRule(
+    public static ConstraintRule constraintRule(
+            long id, ConstraintDescriptor descriptor, String name )
+    {
+        return new ConstraintRule( id, descriptor, null, name );
```
I was thinking to generate the default name here, but whatever. Having it in SchemaRule brings the functionality together, and that's always nice.

fickludd reviewed (approved):
Nice, will merge when green.
